File: 11.container-with-most-water/solution.kt
class Solution {
    fun maxArea(height: IntArray): Int {
        var left = 0
        var right = height.lastIndex
        var maxVolume = 0
        while (left < right) {
            maxVolume = maxOf(maxVolume, minOf(height[left], height[right]) * (right - left))
            if (height[left] < height[right]) {
                left++
            } else {
                right--
            }
        }
        return maxVolume
    }
}

File: 11.container-with-most-water/solution.go
func maxArea(height []int) int {
    volume := 0
    left, right := 0, len(height) - 1
    for left < right {
        volume = max(volume, min(height[left], height[right]) * (right - left))
        if height[left] < height[right] {
            left++
        } else {
            right--
        }
    }
    return volume
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}